### DISCUSSION TITLE
@formily/vue Switch to Vue 3 get compile error on defineComponent

### DISCUSSION THREAD

**Discussion by cetsupport** [alibaba/formily#1933] [Q&A 2.x]
[Version]: Formily 2.0 formily_next branch. Vue version 3.1.5
[Reproduce procedure]:
1. Clone code to local.
2. yarn install.
3. Manual switch Vue version: npx vue-demi-switch 3
4. compile it with yarn build.
Then get following error:

`src/components/ArrayField.ts(13,3): error TS2769: No overload matches this call.
  Overload 1 of 4, '(setup: (props: Readonly<IFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>>, ctx: SetupContext<...>) => object | RenderFunction): DefineComponent<...>', gave the following error.
    Argument of type '{ name: string; props: { name: {}; title: {}; description: {}; value: {}; initialValue: {}; basePath: {}; decorator: ArrayConstructor; component: ArrayConstructor; ... 12 more ...; reactions: (ArrayConstructor | FunctionConstructor)[]; }; setup(props: IFieldProps<...>, { slots }: { ...; }): () => any; }' is not assignable to parameter of type '(props: Readonly<IFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>>, ctx: SetupContext<...>) => object | RenderFunction'.
      Object literal may only specify known properties, and 'name' does not exist in type '(props: Readonly<IFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>>, ctx: SetupContext<...>) => object | RenderFunction'.
  Overload 2 of 4, '(options: ComponentOptionsWithoutProps<IFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>, {}, {}, {}, {}, ComponentOptionsMixin, ComponentOptionsMixin, EmitsOptions, string>): DefineComponent<...>', gave the following error.
    Type '{ name: {}; title: {}; description: {}; value: {}; initialValue: {}; basePath: {}; decorator: ArrayConstructor; component: ArrayConstructor; display: StringConstructor; ... 11 more ...; reactions: (ArrayConstructor | FunctionConstructor)[]; }' is not assignable to type 'undefined'.
src/components/Field.ts(11,16): error TS2769: No overload matches this call.
  Overload 1 of 4, '(setup: (props: Readonly<IFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>>, ctx: SetupContext<...>) => object | RenderFunction): DefineComponent<...>', gave the following error.
    Argument of type '{ name: string; props: { name: {}; title: {}; description: {}; value: {}; initialValue: {}; basePath: {}; decorator: ArrayConstructor; component: ArrayConstructor; ... 12 more ...; reactions: (ArrayConstructor | FunctionConstructor)[]; }; setup(props: IFieldProps<...>, { slots }: { ...; }): () => any; }' is not assignable to parameter of type '(props: Readonly<IFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>>, ctx: SetupContext<...>) => object | RenderFunction'.
      Object literal may only specify known properties, and 'name' does not exist in type '(props: Readonly<IFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>>, ctx: SetupContext<...>) => object | RenderFunction'.
  Overload 2 of 4, '(options: ComponentOptionsWithoutProps<IFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>, {}, {}, {}, {}, ComponentOptionsMixin, ComponentOptionsMixin, EmitsOptions, string>): DefineComponent<...>', gave the following error.
    Type '{ name: {}; title: {}; description: {}; value: {}; initialValue: {}; basePath: {}; decorator: ArrayConstructor; component: ArrayConstructor; display: StringConstructor; ... 11 more ...; reactions: (ArrayConstructor | FunctionConstructor)[]; }' is not assignable to type 'undefined'.
src/components/FormProvider.ts(10,16): error TS2769: No overload matches this call.
  Overload 1 of 4, '(setup: (props: Readonly<IProviderProps>, ctx: SetupContext<EmitsOptions>) => object | RenderFunction): DefineComponent<...>', gave the following error.
    Argument of type '{ name: string; inheritAttrs: boolean; props: { form: { type: ObjectConstructor; required: boolean; }; }; setup(props: IProviderProps, { attrs, slots }: { attrs: any; slots: any; }): () => any; }' is not assignable to parameter of type '(props: Readonly<IProviderProps>, ctx: SetupContext<EmitsOptions>) => object | RenderFunction'.
      Object literal may only specify known properties, and 'name' does not exist in type '(props: Readonly<IProviderProps>, ctx: SetupContext<EmitsOptions>) => object | RenderFunction'.
  Overload 2 of 4, '(options: ComponentOptionsWithoutProps<IProviderProps, {}, {}, {}, {}, ComponentOptionsMixin, ComponentOptionsMixin, EmitsOptions, string>): DefineComponent<...>', gave the following error.
    Type '{ form: { type: ObjectConstructor; required: true; }; }' is not assignable to type 'undefined'.
src/components/ObjectField.ts(13,3): error TS2769: No overload matches this call.
  Overload 1 of 4, '(setup: (props: Readonly<IFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>>, ctx: SetupContext<...>) => object | RenderFunction): DefineComponent<...>', gave the following error.
    Argument of type '{ name: string; props: { name: {}; title: {}; description: {}; value: {}; initialValue: {}; basePath: {}; decorator: ArrayConstructor; component: ArrayConstructor; ... 12 more ...; reactions: (ArrayConstructor | FunctionConstructor)[]; }; setup(props: IFieldProps<...>, { slots }: { ...; }): () => any; }' is not assignable to parameter of type '(props: Readonly<IFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>>, ctx: SetupContext<...>) => object | RenderFunction'.
      Object literal may only specify known properties, and 'name' does not exist in type '(props: Readonly<IFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>>, ctx: SetupContext<...>) => object | RenderFunction'.
  Overload 2 of 4, '(options: ComponentOptionsWithoutProps<IFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>, {}, {}, {}, {}, ComponentOptionsMixin, ComponentOptionsMixin, EmitsOptions, string>): DefineComponent<...>', gave the following error.
    Type '{ name: {}; title: {}; description: {}; value: {}; initialValue: {}; basePath: {}; decorator: ArrayConstructor; component: ArrayConstructor; display: StringConstructor; ... 11 more ...; reactions: (ArrayConstructor | FunctionConstructor)[]; }' is not assignable to type 'undefined'.
src/components/ReactiveField.ts(17,3): error TS2769: No overload matches this call.
  Overload 1 of 4, '(setup: (props: Readonly<IReactiveFieldProps>, ctx: SetupContext<EmitsOptions>) => object | RenderFunction): DefineComponent<...>', gave the following error.
    Argument of type '{ name: string; props: string[]; setup(props: IReactiveFieldProps, { slots }: { slots: any; }): () => any; }' is not assignable to parameter of type '(props: Readonly<IReactiveFieldProps>, ctx: SetupContext<EmitsOptions>) => object | RenderFunction'.
      Object literal may only specify known properties, and 'name' does not exist in type '(props: Readonly<IReactiveFieldProps>, ctx: SetupContext<EmitsOptions>) => object | RenderFunction'.
  Overload 2 of 4, '(options: ComponentOptionsWithoutProps<IReactiveFieldProps, {}, {}, {}, {}, ComponentOptionsMixin, ComponentOptionsMixin, EmitsOptions, string>): DefineComponent<...>', gave the following error.
    Type 'string[]' is not assignable to type 'undefined'.
src/components/RecursionField.ts(33,3): error TS2769: No overload matches this call.
  Overload 1 of 4, '(setup: (props: Readonly<IRecursionFieldProps>, ctx: SetupContext<EmitsOptions>) => object | RenderFunction): DefineComponent<...>', gave the following error.
    Argument of type '{ name: string; inheritAttrs: boolean; props: { schema: { required: boolean; }; name: (StringConstructor | NumberConstructor)[]; basePath: {}; onlyRenderProperties: { ...; }; onlyRenderSelf: { ...; }; mapProperties: {}; filterProperties: {}; }; setup(props: IRecursionFieldProps): () => any; }' is not assignable to parameter of type '(props: Readonly<IRecursionFieldProps>, ctx: SetupContext<EmitsOptions>) => object | RenderFunction'.
      Object literal may only specify known properties, and 'name' does not exist in type '(props: Readonly<IRecursionFieldProps>, ctx: SetupContext<EmitsOptions>) => object | RenderFunction'.
  Overload 2 of 4, '(options: ComponentOptionsWithoutProps<IRecursionFieldProps, {}, {}, {}, {}, ComponentOptionsMixin, ComponentOptionsMixin, EmitsOptions, string>): DefineComponent<...>', gave the following error.
    Type '{ schema: { required: true; }; name: (StringConstructor | NumberConstructor)[]; basePath: {}; onlyRenderProperties: { type: BooleanConstructor; default: undefined; }; onlyRenderSelf: { ...; }; mapProperties: {}; filterProperties: {}; }' is not assignable to type 'undefined'.
src/components/SchemaField.ts(133,23): error TS2769: No overload matches this call.
  Overload 1 of 4, '(setup: (props: Readonly<ISchemaFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>, ObjectField<VueComponent<Record<string, any>>, VueComponent<...>>>>, ctx: SetupContext<...>) => object | RenderFunction): DefineComponent<...>', gave the following error.
    Argument of type '{ name: string; inheritAttrs: boolean; props: { schema: {}; scope: {}; components: {}; basePath: {}; title: {}; description: {}; value: {}; initialValue: {}; validator: {}; dataSource: {}; name: (StringConstructor | NumberConstructor)[]; ... 12 more ...; readPretty: { ...; }; }; setup(props: ISchemaFieldProps<...>, ...' is not assignable to parameter of type '(props: Readonly<ISchemaFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>, ObjectField<VueComponent<Record<string, any>>, VueComponent<...>>>>, ctx: SetupContext<...>) => object | RenderFunction'.
      Object literal may only specify known properties, and 'name' does not exist in type '(props: Readonly<ISchemaFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>, ObjectField<VueComponent<Record<string, any>>, VueComponent<...>>>>, ctx: SetupContext<...>) => object | RenderFunction'.
  Overload 2 of 4, '(options: ComponentOptionsWithoutProps<ISchemaFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>, ObjectField<VueComponent<Record<string, any>>, VueComponent<...>>>, ... 7 more ..., string>): DefineComponent<...>', gave the following error.
    Type '{ schema: {}; scope: {}; components: {}; basePath: {}; title: {}; description: {}; value: {}; initialValue: {}; validator: {}; dataSource: {}; name: (StringConstructor | NumberConstructor)[]; ... 12 more ...; readPretty: { ...; }; }' is not assignable to type 'undefined'.
src/components/SchemaField.ts(255,23): error TS2769: No overload matches this call.
  Overload 1 of 4, '(setup: (props: Readonly<Stringify<{ version?: string; name?: string | number; title?: string; description?: string; default?: any; readOnly?: boolean; writeOnly?: boolean; type?: SchemaTypes; enum?: SchemaEnum<string>; ... 38 more ...; "x-read-pretty"?: boolean; }>>, ctx: SetupContext<...>) => object | RenderFunction): DefineComponent<...>', gave the following error.
    Argument of type '{ name: string; props: { version: StringConstructor; name: (StringConstructor | NumberConstructor)[]; title: {}; description: {}; ... 40 more ...; xReadPretty: { ...; }; } & { ...; }; setup(props: any, { slots }: { ...; }): () => any; }' is not assignable to parameter of type '(props: Readonly<Stringify<{ version?: string; name?: string | number; title?: string; description?: string; default?: any; readOnly?: boolean; writeOnly?: boolean; type?: SchemaTypes; enum?: SchemaEnum<string>; ... 38 more ...; "x-read-pretty"?: boolean; }>>, ctx: SetupContext<...>) => object | RenderFunction'.
      Object literal may only specify known properties, and 'name' does not exist in type '(props: Readonly<Stringify<{ version?: string; name?: string | number; title?: string; description?: string; default?: any; readOnly?: boolean; writeOnly?: boolean; type?: SchemaTypes; enum?: SchemaEnum<string>; ... 38 more ...; "x-read-pretty"?: boolean; }>>, ctx: SetupContext<...>) => object | RenderFunction'.
  Overload 2 of 4, '(options: ComponentOptionsWithoutProps<Stringify<{ version?: string; name?: string | number; title?: string; description?: string; default?: any; readOnly?: boolean; writeOnly?: boolean; type?: SchemaTypes; enum?: SchemaEnum<string>; ... 38 more ...; "x-read-pretty"?: boolean; }>, ... 7 more ..., string>): DefineComponent<...>', gave the following error.
    Type '{ version: StringConstructor; name: (StringConstructor | NumberConstructor)[]; title: {}; description: {}; default: {}; readOnly: { ...; }; ... 38 more ...; xReadPretty: { ...; }; } & { ...; }' is not assignable to type 'undefined'.
src/components/SchemaField.ts(308,12): error TS2769: No overload matches this call.
  Overload 1 of 4, '(setup: (props: Readonly<Stringify<{ version?: string; name?: string | number; title?: string; description?: string; default?: any; readOnly?: boolean; writeOnly?: boolean; type?: SchemaTypes; enum?: SchemaEnum<string>; ... 38 more ...; "x-read-pretty"?: boolean; }>>, ctx: SetupContext<...>) => object | RenderFunction): DefineComponent<...>', gave the following error.
    Argument of type '{ name: string; props: { version: StringConstructor; name: (StringConstructor | NumberConstructor)[]; title: {}; description: {}; ... 40 more ...; xReadPretty: { ...; }; }; setup(props: Stringify<...>, { slots }: { ...; }): () => any; }' is not assignable to parameter of type '(props: Readonly<Stringify<{ version?: string; name?: string | number; title?: string; description?: string; default?: any; readOnly?: boolean; writeOnly?: boolean; type?: SchemaTypes; enum?: SchemaEnum<string>; ... 38 more ...; "x-read-pretty"?: boolean; }>>, ctx: SetupContext<...>) => object | RenderFunction'.
      Object literal may only specify known properties, and 'name' does not exist in type '(props: Readonly<Stringify<{ version?: string; name?: string | number; title?: string; description?: string; default?: any; readOnly?: boolean; writeOnly?: boolean; type?: SchemaTypes; enum?: SchemaEnum<string>; ... 38 more ...; "x-read-pretty"?: boolean; }>>, ctx: SetupContext<...>) => object | RenderFunction'.
  Overload 2 of 4, '(options: ComponentOptionsWithoutProps<Stringify<{ version?: string; name?: string | number; title?: string; description?: string; default?: any; readOnly?: boolean; writeOnly?: boolean; type?: SchemaTypes; enum?: SchemaEnum<string>; ... 38 more ...; "x-read-pretty"?: boolean; }>, ... 7 more ..., string>): DefineComponent<...>', gave the following error.
    Type '{ version: StringConstructor; name: (StringConstructor | NumberConstructor)[]; title: {}; description: {}; default: {}; readOnly: { ...; }; ... 38 more ...; xReadPretty: { ...; }; }' is not assignable to type 'undefined'.
src/components/VoidField.ts(11,16): error TS2769: No overload matches this call.
  Overload 1 of 4, '(setup: (props: Readonly<IVoidFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>>, ctx: SetupContext<...>) => object | RenderFunction): DefineComponent<...>', gave the following error.
    Argument of type '{ name: string; props: { name: {}; title: {}; description: {}; basePath: {}; decorator: ArrayConstructor; component: ArrayConstructor; display: StringConstructor; ... 7 more ...; reactions: (ArrayConstructor | FunctionConstructor)[]; }; setup(props: IVoidFieldProps<...>, { slots }: { ...; }): () => any; }' is not assignable to parameter of type '(props: Readonly<IVoidFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>>, ctx: SetupContext<...>) => object | RenderFunction'.
      Object literal may only specify known properties, and 'name' does not exist in type '(props: Readonly<IVoidFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>>, ctx: SetupContext<...>) => object | RenderFunction'.
  Overload 2 of 4, '(options: ComponentOptionsWithoutProps<IVoidFieldProps<VueComponent<Record<string, any>>, VueComponent<Record<string, any>>>, {}, {}, {}, {}, ComponentOptionsMixin, ComponentOptionsMixin, EmitsOptions, string>): DefineComponent<...>', gave the following error.
    Type '{ name: {}; title: {}; description: {}; basePath: {}; decorator: ArrayConstructor; component: ArrayConstructor; display: StringConstructor; ... 7 more ...; reactions: (ArrayConstructor | FunctionConstructor)[]; }' is not assignable to type 'undefined'.
src/vue2-components.ts(4,15): error TS2305: Module '"../../../node_modules/vue/dist/vue"' has no exported member 'VueConstructor'.`

Question:
Do I need manual change the Vue version or just compile it afterward change in project?

Thank for help

**Comment by MisicDemone** [comment]
Needn't to switch to Vue 3. Just run `yarn build`.
